src/compiler/checker.ts

```diff
@@ -18514,10 +18514,20 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return !!(getUnionType([intersectTypes(type1, type2), neverType]).flags & TypeFlags.Never);
     }
 
+    function hasOptionalModifiersProperty(mappedType: MappedType) {
+        const modifiersType = getApparentType(getModifiersTypeFromMappedType(mappedType));
+        if (!(modifiersType.flags & TypeFlags.Object)) {
+            return false;
+        }
+        const properties = resolveStructuredTypeMembers(modifiersType as ObjectType).properties;
+        return some(properties, p => !!(p.flags & SymbolFlags.Optional));
+    }
+
     function substituteIndexedMappedType(objectType: MappedType, index: Type) {
         const mapper = createTypeMapper([getTypeParameterFromMappedType(objectType)], [index]);
         const templateMapper = combineTypeMappers(objectType.mapper, mapper);
-        return instantiateType(getTemplateTypeFromMappedType(objectType.target as MappedType || objectType), templateMapper);
+        const type = instantiateType(getTemplateTypeFromMappedType(objectType.target as MappedType || objectType), templateMapper);
+        return addOptionality(type, /*isProperty*/ true, hasOptionalModifiersProperty(objectType));

```

Comment:
TODO: this should get the list of known optional properties and check if `index` is assignable to that, if the substitution happens using a known non-optional property then the optionality shouldn't be added